### DISCUSSION TITLE
Support Jamulus as VST plugin

### DISCUSSION THREAD

**Discussion by corrados** [jamulussoftware/jamulus#2458] [Ideas]
_This issue has been migrated from Sourceforge.  Created: 2010-01-27
Creator: Volker Fischer_

The title says everything...

Discussion

    Volker Fischer - 2010-01-31

This one might help implementation:
http://ag-works.net/source.wizards.htm

**Comment by vocobox** [comment]
It may be clarified. 
- I presume it means that Jamulus should appear as input to other software, such as Cubase etc. This would allow to record a Jamulus session to a DAW locally.
- Or should it be able to ingest other software's output so that we can stream any audio software output to a Jamulus server? This would allow to stream a local multitrack session to a Jamulus server.

Both are very valuable! 

I refer to existing forum discussion 
- one of your note concerning your [attempt to integrate VST](https://sourceforge.net/p/llcon/discussion/533517/thread/31a9827855/?limit=25)
- a forum member note on the ability to [use JACK to interact with Jamulus](https://sourceforge.net/p/llcon/discussion/533517/thread/bb6951c1/)

**Comment by trebmuh** [comment]
We can already achieve both with jack. What OS do you use?

**Comment by vocobox** [comment]
I have MacOS 10.15.3 for programming, 10.12.3 for audio work. Despite I could not install Jack on 10.15.3, I can install it on 10.12.3 and when started, Jamulus displays Jack in the list of devices. 

Beside that, when I start my virtual synths (Arturia) once Jack server is started I often go crashes of Jack server saying that the sampling freq was changed or buffer size can't be changed, but that's another topic I'll probably understand in Jack's user manual. 

Thank you @trebmuh

**Comment by Snayler** [comment]
Some information regarding this:

pljones mentioned in the Facebook group that
> The issue with the VST protocol up to version 2 is that the license prevents Jamulus adopting it.

While this is true, there is a workaround. As discussed [here](https://github.com/rorywalsh/cabbage_v1_old/issues/5), one could develop the VST plugin support but ask users to compile it themselves, as it would not go against Steinberg's License. So the normal binaries wouldn't have VST support, but instructions could be given so that people can compile Jamulus with VST support themselves.

Another alternative would be developing a Audio Units plugin, as it's license [is compatible with GPL](http://virtualanalogy.com/gnu-gpl-and-audio-plugins-iii-license-compatibility-ii/).

**Comment by corrados** [comment]
For the records: The initial CR was from myself and it was created in 2010 (10 years ago :-) ). Back in the day I already started some implementation:
https://github.com/corrados/jamulus/blob/master/src/vstmain.cpp
https://github.com/corrados/jamulus/blob/master/src/vstmain.h
https://github.com/corrados/jamulus/blob/master/src/vstsound.h
EDIT by @pljones the work is now here: https://github.com/jamulussoftware/jamulus/tree/archive/corrados/vst
As far as I remember, something might already work. But there is still a lot to do to make it "really" work. Anyone interested in taking over that work (i.e. finish that work)?

**Reply by yatli** [reply]
for anyone interested, the code is not accessible from latest master but still accessible in the vst feature branch: https://github.com/jamulussoftware/jamulus/tree/feature/vst/src

**Reply by pljones** [reply]
The work is now here: https://github.com/jamulussoftware/jamulus/tree/archive/corrados/vst

**Comment by trebmuh** [comment]
Note that in case of working on this, a LV2 version would be nice too. Maybe [the DPF plugin framework](https://github.com/DISTRHO/DPF) can be of help here.

**Comment by WolfganP** [comment]
I'm sorry, but what will be the use case for the VST? 
Get the server mix into a DAW (which is already achievable) or the client to feed individual streams to a local DAW? 
Or the other way around (DAW output as a channel to stream via jamulus client) which is also achievable right now via ASIO routing?

**Comment by Snayler** [comment]
> I'm sorry, but what will be the use case for the VST?

One possible use would be using a different DAW than Reaper to get the server mix. Sure, one could load all the tracks into another DAW, but the tracks would have to be aligned by hand, or you'd have to spend money on a DAW project converter (last time I researched this they were quite costly). That's the only valid use-case I can come up with, as routing the sound is already available through ReaRouter, Jack, VB-Audio, etc...

**Comment by WolfganP** [comment]
> > I'm sorry, but what will be the use case for the VST?
> 
> One possible use would be using a different DAW than Reaper to get the server mix. Sure, one could load all the tracks into another DAW, but the tracks would have to be aligned by hand, or you'd have to spend money on a DAW project converter (last time I researched this they were quite costly). That's the only valid use-case I can come up with, as routing the sound is already available through ReaRouter, Jack, VB-Audio, etc...

I'm sorry, but still not getting it. Are you talking about online/streaming DAW usage in the chain or processing of recordings? None of them makes Reaper a mandatory or exclusive app.

**Comment by Snayler** [comment]
I'm talking purely about recordings. The server creates the recording and the Reaper Project File. Imagine you don't want to use Reaper, but instead want to use something like, let's say. Pro Tools. You'd have to load the tracks into a new project and adjust the timings of each track by hand. On long recordings this could be very time consuming and having a plugin that enables any DAW to record all the tracks on the fly would be a plus.

I'm personally happy with Reaper, but this is what I gathered from the discussion going on the Facebook group.

**Comment by WolfganP** [comment]
Now I understand, but to be honest, I don't imagine how a Jamulus as VST will help you on that matter. 

The .rpp file is a text one, so it's easy to keep it open in a separate window to adjust the start time for each track (in a typical band jam scenario most of the time you'll end up with less than 10 tracks per session, so it won't be too much work to add to all the stuff you plan to do with the recording on the DAW).

**Comment by Snayler** [comment]
> The .rpp file is a text one, so it's easy to keep it open in a separate window to adjust the start time for each track (in a typical band jam scenario most of the time you'll end up with less than 10 tracks per session, so it won't be too much work to add to all the stuff you plan to do with the recording on the DAW).

Well, I have a 5 hour recording with people coming in and out, 20 channels after cleaning the channels that don't matter. 30 tracks total (since some people disconnected and reconnected later during the session).

>I don't imagine how a Jamulus as VST will help you on that matter.

My idea would be something like ReaRouter, but on Jamulus server. It would have multiple virtual audio channels that a user could assign to multiple tracks on the preferred DAW and record a session.

**Comment by WolfganP** [comment]
> Well, I have a 5 hour recording with people coming in and out, 20 channels after cleaning the channels that don't matter. 30 tracks total (since some people disconnected and reconnected later during the session).

I deleted/converted all my recordings to check the file naming schema, but if not sufficiently descriptive that maybe introduced as a feature request for being able to script the import to any DAW around (ie DANNY-20200525T213304.157.wav or something similar)
@pljones posted interesting recording process scripts at https://github.com/corrados/jamulus/issues/206#issuecomment-627437159
 
> My idea would be something like ReaRouter, but on Jamulus server. It would have multiple virtual audio channels that a user could assign to multiple tracks on the preferred DAW and record a session.

The only place where all the client streams are separated are at the server side, a similar request was recently discussed at https://github.com/corrados/jamulus/issues/146 so you may be interested in reading that issue

**Comment by Snayler** [comment]
> @pljones posted interesting recording process scripts at [#206 (comment)](https://github.com/corrados/jamulus/issues/206#issuecomment-627437159)

That's quite interesting.

Anyway, I'm happy with Reaper, I recently bought it as I was surprised such a good DAW was available so cheap (when compared with Cubase which is what I had before). There are some people on the [Facebook group](https://www.facebook.com/groups/619274602254947/permalink/664437854405288/) that are interested in this.

**Comment by gilgongo** [comment]
Some people might be interested in https://github.com/jamulussoftware/jamulus/issues/874

**Comment by ann0see** [comment]
Converted to discussion as this isn't been worked on actively.
The work on VST is now here: https://github.com/jamulussoftware/jamulus/tree/feature/vst
Edit: @pljones -- The work is now here: https://github.com/jamulussoftware/jamulus/tree/archive/corrados/vst

**Comment by baboo35** [comment]
I’ve tested the SonoBus LV2 plugin in Reaper yesterday, and it works flawlessly. The ability to record a remote realtime jam session based on my working tracks with just one click directly inside my favorite DAW — is a great feature.

This experience made me think that Jamulus should follow a similar path by developing a VST, LV2, or similar plugin for DAWs.

And of course, such integration would significantly expand Jamulus’s user base:)

**Reply by pljones** [reply]
Recording would still be the lossy stereo return from the server, of course, so no gain over using PipeWire or JACK.
